### ISSUE TITLE
Error World War 2 Europe 1940 Original

### ISSUE THREAD

**Issue by ZjelcoP** [triplea-game/triplea#3100]
WARNING: Could not set property:turns subject:RulesAttachment attached to:PlayerID 

named:French with name:conditionAttachment_French_1_Liberation new value:1-+

### My Operating System:
Windows 7
### Engine version:
9065 and 9106
### Map name and version:
World War 2 Europe 1940 Original
### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)
Ending of German turn. (Set Germany as AI, let it run)
### Do you have the exact error text? Please copy/paste if so
[WW2 Europe 1940 Original.txt](https://github.com/triplea-game/triplea/files/1745617/WW2.Europe.1940.Original.txt)
### Any additional information that may help: 
Reproduced twice in both 9065 and 9106.
7621 no error.

**Comment by ssoloff** [comment]
So, it appears the `AbstractRulesAttachment#setTurns(HashMap<Integer, Integer>)` method is being called with an argument of type `String`.  Guess I'm wondering why there's no overload `setTurns(String)` like most other attachment setters have.  Anyway, I don't see any recent changes, for example, where such an overload was removed.  Will need to keep digging.

**Comment by RoiEXLab** [comment]
@ssoloff My PR adds a String setter for all attachments I think, so that might fix the issue already...

**Comment by ssoloff** [comment]
Actually, it looks like the intention is to call `AbstractRulesAttachment#setRounds(String)`.  That appears to have the logic to set `m_turns` using the specific string format present in the error log.  So now the question is why is the property name `turns` instead of `rounds`?

**Comment by ssoloff** [comment]
@RoiEXLab BTW, I think this code demonstrates that attachments might not be immutable, as we theorized in #3042.  The fact that there's a `ChangeAttachmentChange` class is pretty good evidence attachment properties can change during a game.

**Comment by RoiEXLab** [comment]
@ssoloff Even if there are a couple of mutable fields, I still think it would make sense to initialize them at once and make as much variables final as possible.

**Comment by ssoloff** [comment]
Note that I can reproduce the issue with a 7621 server and a 3635 client, which suggests the issue is not relatively new.

I looked at the _ww2eur40.xml_ map XML and found that it still contains two `turns` properties:

https://github.com/triplea-maps/world_war_ii_europe/blob/d71a8ef28d9aa7d4c6dca0af779cea3d10f2689b/map/games/ww2eur40.xml#L2347-L2363

```xml
    <!-- French Triggers -->
    <attachment name="triggerAttachment_Germans_Conquer_France" attachTo="Germans" javaClass="games.strategy.triplea.attachments.TriggerAttachment" type="player">
      <option name="trigger" value="conditionAttachment_French_1_Lose_France"/>
      <option name="uses" value="1"/>
      <option name="when" value="after:germansBattle"/>
      <option name="players" value="French"/>
      <option name="playerAttachmentName" value="RulesAttachment" count="conditionAttachment_French_1_Liberation"/>
      <option name="playerProperty" value="turns" count="1-+"/>
    </attachment>
    <attachment name="triggerAttachment_Italians_Conquer_France" attachTo="Italians" javaClass="games.strategy.triplea.attachments.TriggerAttachment" type="player">
      <option name="trigger" value="conditionAttachment_French_1_Lose_France"/>
      <option name="uses" value="1"/>
      <option name="when" value="after:italiansBattle"/>
      <option name="players" value="French"/>
      <option name="playerAttachmentName" value="RulesAttachment" count="conditionAttachment_French_1_Liberation"/>
      <option name="playerProperty" value="turns" count="1-+"/>
    </attachment>
```

My guess is these `turns` properties were missed when they were supposed to be bulk replaced to `rounds` in [7d78f75](https://github.com/triplea-maps/world_war_ii_europe/commit/7d78f75da3c11b01743bdc09665a937d44b74ca9#diff-ec2206525ffe9f1640087cf29014b7d7).

**Comment by ssoloff** [comment]
Just a repro update: playing a network game is not necessary.  A local game in which Germany conquers France will also trigger the error.

**Comment by ssoloff** [comment]
Changing `turns` to `rounds` as the value of the `playerProperty` attachment option in the above XML fixes the original error but generates a different one:

```
Error: No such Property Field named: m_rounds, or: rounds, for Subject: RulesAttachment attached to:PlayerID named:French with name:conditionAttachment_French_1_Liberation
java.lang.IllegalStateException: No such Property Field named: m_rounds, or: rounds, for Subject: RulesAttachment attached to:PlayerID named:French with name:conditionAttachment_French_1_Liberation
	at games.strategy.util.PropertyUtil.getPropertyFieldObject(PropertyUtil.java:78)
	at games.strategy.engine.data.DefaultAttachment.getRawPropertyString(DefaultAttachment.java:113)
	at games.strategy.triplea.attachments.TriggerAttachment.triggerPlayerPropertyChange(TriggerAttachment.java:1629)
	at games.strategy.triplea.attachments.TriggerAttachment.fireTriggers(TriggerAttachment.java:222)
	at games.strategy.triplea.attachments.TriggerAttachment.collectAndFireTriggers(TriggerAttachment.java:168)
	at games.strategy.triplea.delegate.BaseTripleADelegate.triggerWhenTriggerAttachments(BaseTripleADelegate.java:90)
	at games.strategy.triplea.delegate.BaseTripleADelegate.end(BaseTripleADelegate.java:59)
	at games.strategy.triplea.delegate.BattleDelegate.end(BattleDelegate.java:119)
	at games.strategy.engine.framework.ServerGame.endStep(ServerGame.java:454)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:418)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:266)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInNewThread(LocalLauncher.java:58)
	at games.strategy.engine.framework.startup.launcher.AbstractLauncher.lambda$0(AbstractLauncher.java:51)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: No such Property Field: rounds
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:60)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:65)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:65)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:65)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:65)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:65)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:65)
	at games.strategy.util.PropertyUtil.getFieldIncludingFromSuperClasses(PropertyUtil.java:55)
	at games.strategy.util.PropertyUtil.getPropertyField(PropertyUtil.java:103)
	at games.strategy.util.PropertyUtil.getPropertyField(PropertyUtil.java:83)
	at games.strategy.util.PropertyUtil.getPropertyFieldObject(PropertyUtil.java:71)
	... 13 more
```

I need to step away for a few hours, but I'll look into it again later this evening.  Any suggestions for the correct fix are welcome.

**Comment by ssoloff** [comment]
@ZjelcoP Thanks for the link.  I thought this error looked familiar.  It's sad that I can't make a connection from 12 days ago. :smile:

I did a little more debugging:

* I was able to reproduce the problem in 3635 (although the console doesn't automatically open; you have to manually open it to see the error).  I believe the root cause goes all the way back to 0cd3154a5 when `AbstractRuleAttachment#setTurns(String)` was renamed to `setRounds()`.
* As I said above, simply changing the `playerProperty` value in the XML to `rounds` doesn't work because `rounds` is not a true property, but rather what we've termed a _virtual_ property.  It appears that only some of the reflection code works with virtual properties (e.g. `ChangeAttachmentChange`) and some doesn't (e.g. `TriggerAttachment#triggerPlayerPropertyChange()`).
* I tried restoring the `AbstractRuleAttachment#setTurns(String)` method (and simply have it pass through to `setRounds(String)`).  That appears to fix the problem without any map changes.  I'll submit a PR shortly so this possible solution can be discussed by the dev team.

**Comment by ssoloff** [comment]
@ZjelcoP This should be fixed in 1.9.0.0.9131 or later.
